### PR TITLE
Update tfjob launcher

### PR DESCRIPTION
**Pull request by hougangliu**
the tfjob_launcher is out of date to support tfjob v1beta

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1494)

### CODE REVIEW

**Comment by ryandawsonuk** [comment]
/lgtm

I've raised a question about pvolumes for this in https://github.com/kubeflow/pipelines/issues/1344#issuecomment-503471782 but I don't personally think that should prevent this being merged

File: components/kubeflow/launcher/kubeflow_tfjob_launcher_op.py [hunk 1]
```diff
@@ -14,17 +14,31 @@
...
+                               has_master: bool,
+                               number_of_workers: int,
+                               number_of_parameter_servers: int,
+                               tfjob_timeout_minutes: int,
+                               output_dir=None,
+                               step_name='TFJob-launcher'):
     return dsl.ContainerOp(
         name = step_name,
-        image = 'gcr.io/ml-pipeline/ml-pipeline-kubeflow-tf:1d55a27cf8b69696f3ab5c10687edf2fde0068c7',
+        image = 'liuhougangxa/tfjob-launcher',
```
Comment (ryandawsonuk) [inline components/kubeflow/launcher/kubeflow_tfjob_launcher_op.py]: Is this intentional?

**Comment by elikatsis** [comment]
@hougangliu, Hello!

I would like to suggest some other type of mapping.
Instead of a dict `{"pvc-name": "/mount/path"}`, I believe it would be better to use `{"/mount/path": some_k8s_volume_instance}`.

Then, the following lines
https://github.com/kubeflow/pipelines/blob/4aca05e1d41941364ad54449ef443444deebc724/components/kubeflow/launcher/src/launch_tf_job.py#L70-L72

will become
```python
for k, v in vol_map.iteritems()
    spec['volumes] = v
    spec['containers'][0]['volumeMounts'] = [{"mountPath": k, "name": v.name}]
```

In addition, it will allow the mounting of other types of volumes (e.g. `Secret`s).

This will also be uniform with the argument `pvolumes` of `ContainerOp` entity.
[ For more info check the `volumeop` samples [here](https://github.com/kubeflow/pipelines/tree/master/samples/resourceops), and also [these lines](https://github.com/kubeflow/pipelines/blob/79411b7bbaae288bc48cbf79d5006d836080483f/sdk/python/kfp/dsl/_container_op.py#L927-L929) of `ContainerOp` definition. ]

It can then be parsed accordingly to extend the dependencies of that task, in the case where a `PipelineVolume` instance is passed as a value.

I'll be holding the PR until we have discussed that thoroughly.
/hold

File: components/kubeflow/launcher/kubeflow_tfjob_launcher_op.py (hunk 1 above)
Comment (SatwikBhandiwad) [inline components/kubeflow/launcher/kubeflow_tfjob_launcher_op.py]: is this code different from gcr.io/ml-pipeline/ml-pipeline-kubeflow-tf:1d55a27cf8b69696f3ab5c10687edf2fde0068c7 ? if yes, can we see source code of this docker image?

Comment (hougangliu) [reply components/kubeflow/launcher/kubeflow_tfjob_launcher_op.py]: yes, in fact kubeflow_tfjob_launcher_op.py is just used as an example about what tfjob-launcher  containerOP should be
when image updated in gcr.io/ml-pipeline, the image will be updated, too

Comment (hougangliu) [reply components/kubeflow/launcher/kubeflow_tfjob_launcher_op.py]: components/kubeflow/launcher/src/launch_tf_job.py is the source code

File: components/kubeflow/launcher/src/launch_tf_job.py [hunk 2]
```diff
@@ -41,65 +43,90 @@
...
-                           'cluster is used.')
-  parser.add_argument('--zone', type=str, help='zone of the kubeflow cluster.')
-  parser.add_argument('--kfversion', type=str,
-                      default='v1alpha2',
-                      help='The version of the deployed kubeflow. ' +
-                           'If not set, the default version is v1alpha2')
+  parser.add_argument('--tfjob-version', type=str,
+                      default='v1beta2',
+                      help='The version of the deployed tfjob.' +
+                           'If not set, the default namespace is v1beta2.')
```
Comment (nrchakradhar) [inline components/kubeflow/launcher/src/launch_tf_job.py]: NIT: I guess there is typo here about **namespace**. It should have been **tfjob-version**

File: components/kubeflow/launcher/src/launch_tf_job.py [hunk 3]
```diff
@@ -41,65 +43,90 @@
 import six
 import time
 import yaml
-from py import tf_job_client
+from kubeflow.tf_operator import tf_job_client
 from kubernetes import client as k8s_client
 from kubernetes import config
 
+def yamlOrJsonStr(str):
```
Comment (Ark-kun) [inline components/kubeflow/launcher/src/launch_tf_job.py]: YAML is superset of JSON, so you can just use YAML loader.

File: components/kubeflow/launcher/Dockerfile [hunk 4]
```diff
@@ -47,16 +45,18 @@ RUN wget -nv https://github.com/ksonnet/ksonnet/releases/download/v0.9.0/ks_0.9.
 RUN wget https://github.com/kubeflow/tf-operator/archive/v0.3.0.zip && \
     unzip v0.3.0.zip && \
     mv tf-operator-0.3.0 tf-operator
+RUN wget https://github.com/kubeflow/tf-operator/archive/master.zip && \
```
Comment (Ark-kun) [inline components/kubeflow/launcher/Dockerfile]: Pinning the version is usually preferred to using master version.

**Comment by Ark-kun** [comment]
> May I hold this PR to discuss it a bit longer? I want to ensure that we don't merge something that introduces a different API (since one already exists), which might need refactoring later on to support more functionalities. What do you think?
> /hold

@elikatsis Do you think we should release the hold? Usually the holds are pretty short.

**Comment by hougangliu** [comment]
I will update this PR later today to solve the comment. Sorry for late action

**Comment by ryandawsonuk** [comment]
Worth noting v1beta is [no longer the TFJob version as it's now v1](https://github.com/kubeflow/pipelines/issues/2615)

**Comment by hougangliu** [comment]
#2677 replace this PR

**Pull request closed without merging**